### DISCUSSION TITLE
Thanks! (And pretty cool!)

### DISCUSSION THREAD

**Discussion by nnshah1** [pupnp/pupnp#245] [General]
I wanted to pop in and quickly thank the current maintainers for their work on this project and for keeping it supported for over 20 years! It's quite gratifying to see it continue to be useful and out in the wild. I was one of the original developers back in 1999 - 2001 at Intel (specifically I worked on the threadutils and gena support - and remember being quite happy with UpnpPrintf). This was my first project and our small team put a lot of energy, late nights, and passion into it. Even now when we cross paths we become nostalgic for those days, that team, and this project. Anyway - thanks for the work in keeping it alive and glad it was useful in some small way. Reading the issues and discussion around coding style made me chuckle - yes indeed - please forgive us. We were young and we meant no harm. Interestingly - as I've been cleaning out my home office during the pandemic I came across a document I wrote detailing the design of a "Universal UPnP Control Point Library" - *sigh* that library is sadly lost to history ... Good Luck!

**Comment by whyman** [comment]
@nnshah1 Do you have any suggestion or insight on what could cause the threadpool queues to fill? We still get reports of the `100` job limit being hit

**Reply by nnshah1** [reply]
I probably should not have come clean about the parts I worked on :)

I haven't thought about this in a long time, but I would actually suspect gena. spdp is multicast, soap is limited in number of connections/jobs you would create, only gena stands out in my mind for something that could cause the pool to fill up in some bad network state (maybe a control point that subscribed then went away?)

Do you have any more details on the scenarios under which it happens?

**Reply by mrjimenez** [reply]
> I probably should not have come clean about the parts I worked on :)

Lol, now you are doomed! :laughing:

**Reply by ffontaine** [reply]
Concerning the threadpool queue issue, it is indeed probably linked to GENA and Control Points which are subscribing and disconnecting. I added the `--disable-blocking-tcp-connections` ten years ago with 32cffb5bb55a650b1eb962c6fe2e58e6bf4fe2c5 however it is not enabled by default as it could have side effects. The default timeout has also been set to 5 seconds but it is not enough and trying to send GENA events to disconnected CPs will still raise issues. Ideally, we should detect that the CP is disconnected and remove its subsbcriptions but this is not fully compliant with the UPnP Device Architecture which specifies that:

> If  a  subscriber does  not  respond  within  30  seconds,  or  if  the  publisher  is  unable  to  connect  to  the  subscription  URL,  the  publisher  should abandon  sending  this  message  to  the  subscriber  but  shall keep the subscription active and send future event messages to the subscriber until the subscription  expires  or  is  cancelled.

**Comment by mrjimenez** [comment]
Hi Neelay!

I'm really glad you came along, I can imagine the effort you have put into creating this library.

The home page (which by the way is also the README) of this project has tried to put some historical context, but we are truly limited by the fact that there is no original developer in the project. I have no idea how the project ended in Sourceforge and I have no good idea of how the forks happened. All I know is that I have sent one or two patches fixing some small problems and then I got invited to join the project. Since then I have been trying to keep it working and fixing bugs. If you find you can add interesting facts of curiosities about the project, you are more than welcome to do that.

Well, it goes without saying that your patches to code are also more than welcome!

Best regards,
Marcelo.

**Reply by nnshah1** [reply]
Thanks! I'll reach to some of the other original developers and see if at least they'd be willing to be associated with the project :)

In terms of how it moved from one repository to the other - I'm afraid I don't have a lot of context. I actually left Intel in 2002 to attend grad school and then returned in 2006 (when Intel was no longer actively maintaining it).

It's so difficult now to think of how nascent everything was back then - we had aspirations of smart thermostats, fridges, coffee pots - but UPnP really thrived with respect to AV. We even published a white paper (abanonded patent) on controlling devices from outside the home: https://patents.google.com/patent/US20030051009A1/en. 

And of course there is Jack Weast's book: https://www.amazon.com/UPnP-Design-Example-Developers-Universal/dp/0971786119. 

Jack is now an Intel Fellow!

Essentially we created the SDK along side the UPnP specification itself to seed implementations and help push plugfests forward. 

I don't have a lot of time but will see what I can do. 

Perhaps I can ask a basic question: I assume there are multiple complete  rewrites of a linux UPnP library over the years- why maintain this one?

**Reply by whyman** [reply]
The number upnp libraries is actually pretty low. GNOME folks have `gupnp` which is okay if you are already using glib and friends, but other than that much is out there.

Mostly there are lots of stubs that support only `IGD` functionality, so you have not much option if you are trying to write a fully featured UPnP software.

**Reply by mrjimenez** [reply]
I think that many small and embedded projects use it, and it is also included in the [Buildroot project](https://buildroot.org), I have seen some routers using it on the server side. On the client side, [Gerbera](https://github.com/gerbera/gerbera) is probably the greatest source for finding the problems in this library due to their large user base.

So, yes, it is still pretty much used, thank you very much for all the effort you have put into it 20 years ago!
